pandas/core/generic.py

```diff
@@ -2005,10 +2006,13 @@ def empty(self) -> bool_t:
     def __array__(self, dtype: npt.DTypeLike | None = None) -> np.ndarray:
         values = self._values
         arr = np.asarray(values, dtype=dtype)
-        if arr is values and using_copy_on_write():
-            # TODO(CoW) also properly handle extension dtypes
-            arr = arr.view()
-            arr.flags.writeable = False
+        if arr is values and using_copy_on_write() and self._mgr.is_single_block:
```

Comment:
The `arr is values` might prevent catching some EA cases, since it seems that `_values` can still return an EA (so the EA->ndarray conversion only happens in `arr = np.asarray(values ..)`). 

Example (running with this PR):

```python
In [1]: pd.options.mode.copy_on_write = True

In [2]: df = pd.DataFrame({"a": pd.date_range("2012-01-01", periods=3)})

In [3]: arr = np.asarray(df)

In [4]: arr.flags.writeable
Out[4]: True

In [5]: arr[0] = 0

In [6]: df
Out[6]: 
           a
0 1970-01-01
1 2012-01-02
2 2012-01-03
```


For series you left out this check, so maybe can be done here as well?

Reply:
good point, thx